FileIOUtils.cpp

```diff
@@ -0,0 +1,426 @@
+    memset(tmppaldata, 0, 32);
+    memcpy(tmppaldata, tmppalettedata.data(), qMin(32, tmppalettedata.size()));
+    ROMUtils::LoadPalette(&tmppalette, tmppaldata, true);
+    delete[] tmppaldata;
+
+    // Get transparent color id in the palette
+    int transparentcolorId = 0;
+    SelectColorDialog scdialog;
+    scdialog.SetPalette(tmppalette);
+    scdialog.SetColor(0);
+    if(scdialog.exec() == QDialog::Accepted)
+    {
+        transparentcolorId = scdialog.GetSelectedColorId();
+    } else {
+        return false;
+    }
+
+    // transparent-substitute color replacement and load palette
+    tmppalette[transparentcolorId] = 0;
+
+    // nybble exchange not needed
+    // reset bytearray according to the palette bin file
+    for(int i = 0; i != 16; ++i)
+    {
+        char count = 0;
+
+        // Find if the color[i] is in the current palette
+        while(1)
+        {
+            if(tmppalette[i] == ref_palette[count])
+            {
+                break;
+            }
+            ++count;
+            if(count == 16)
+            {
+                if((tmppalette[i] != 0xFF000000) && (tmppalette[i] != 0xFFFFFFFF) && (tmppalette[i] != 0))
+                {
+                    QMessageBox::critical(parent, QObject::tr("Error"), QObject::tr("Palette not suitable!"));
+                    return false;
+                }
+                else if(tmppalette[i] == 0xFF000000) // black
+                {
+                    auto iter = std::find_if(ref_palette.begin(),
+                                             ref_palette.end(), [&](const QRgb& value) {
+                                    return value == tmppalette[i]; });
+                    if (tmppalette.end() != iter) {
+                        count = iter - tmppalette.begin();
+                    } else {
+                    count = 0;
+                    }
+                    break;
+                }
+                else if(tmppalette[i] == 0xFFFFFFFF) // white
+                {
+                    auto iter = std::find_if(ref_palette.begin(),
+                                             ref_palette.end(), [&](const QRgb& value) {
+                                    return value == tmppalette[i]; });
+                    if (tmppalette.end() != iter) {
+                        count = iter - tmppalette.begin();
+                    } else {
+                    count = 0;
+                    }
+                    break;
+                }
+                else if(tmppalette[i] == 0) // transparent
+                {
+                    count = 0;
+                    break;
+                }
+            }
+        }
+        if(transparentcolorId == i)
+        {
+            count = 0;
+        }
+
+        // replace the color[i] in tiledata with the correct id
+        for(int j = 0; j < tmptile8x8data.size(); ++j) // TODO: bugfix here
+        {
+            char tmpchr = tmptile8x8data[j];
+            char l4b, h4b;
```

Comment:
let's call these lower4bits and upper4bits

Reply:
done.